### PR TITLE
SW-609_automate_release_process

### PR DESCRIPTION
**Pull request by TheKeymaster**

### CODE REVIEW

**Comment by zaifastafa** [comment]
@TheKeymaster In the composer json description there seems to be some left over text. Wasn't sure if its intended or not so didn't create a commit.

File: bin/build-release [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+    'FinSearch/phpcs.xml',
+    'FinSearch/phpunit.legacy.xml.dist',
+    'FinSearch/phpunit.xml.dist',
+    'FinSearch/tests/\*',
+    '*.zip',
+    '*.phar',
+];
+
+$logger = new Logger('build-release');
+$formatter = new LineFormatter(null, null, false, true);
```
Comment (dblaichinger) [inline bin/build-release:29]: Why not use the same `ColoredLineFormatter` here?

File: bin/build-release [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+$logger->info('Copying files...');
+exec(sprintf('cp -rf . %s', TMP_DIR));
+chdir(TMP_DIR);
+
+$logger->info('Installing dependencies (may remove dev dependencies)...');
+exec('composer install --no-dev --optimize-autoloader --prefer-dist');
+
+$logger->info('Adding shopware/core and shopware/storefront to the composer.json.');
+$composerJsonData['require']['shopware/core'] = '^6.2||^6.3||^6.4';
+$composerJsonData['require']['shopware/storefront'] = '^6.2||^6.3||^6.4';
```
Comment (dblaichinger) [inline bin/build-release:61]: Those version might need an update in future, how do we not forget it?

File: bin/build-release [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+file_put_contents(
+    TMP_DIR . '/composer.json',
+    json_encode($composerJsonData, JSON_PRETTY_PRINT|JSON_UNESCAPED_SLASHES|JSON_UNESCAPED_UNICODE) . "\n"
```
Comment (dblaichinger) [inline bin/build-release:64]: `$jsonWithTwoSpaces = preg_replace('/^( +?)\\1(?=[^ ])/m', '$1', $rawJson);` is not done here, is this correct?

File: bin/build-release [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+);
+
+chdir(SYS_TMP_DIR);
+
+$zipFileDest = sprintf('%s/FinSearch.zip', __DIR__ . '/..');
+exec(sprintf('zip -r9 %s FinSearch -x %s', $zipFileDest, implode(' ', IGNORED)));
+
+exec('rm -rf "/tmp/FinSearch"');
```
Comment (dblaichinger) [inline bin/build-release:72]: Shouldn't `SYS_TMP_DIR` be used here as well?

File: bin/build-release [hunk 5]
```diff
@@ -0,0 +1,82 @@
+#!/usr/bin/env php
+<?php
+
```
Comment (dblaichinger) [inline bin/build-release:3]: `declare(strict_types=1);` not used here

File: bin/release [hunk 6]
```diff
@@ -1,81 +1,92 @@
...
 }
 
-$logger->info(sprintf('SUCCESS! The release was successfully created! You can find it in %s', realpath($zipFileDest)));
+function pushVersionChange(LoggerInterface $logger, string $version): void
+{
+    $logger->info('Committing and pushing the version change...');
+
+    exec('git reset HEAD -- .');
+    exec('git add composer.json composer.lock');
+    exec('git commit -m "Bump version"');
```
Comment (dblaichinger) [inline bin/release:85]: When `$version` is available please add it to the commit msg.

File: composer.json [hunk 7]
```diff
@@ -1,7 +1,7 @@
 {
   "name": "findologic/plugin-shopware-6",
   "description": "FINDOLOGIC plugin for Shopware 6 e-commerce system",
-  "version": "v2.1.2",
+  "version": "2.2.0",
```
Comment (dblaichinger) [inline composer.json:4]: Version was with `v` before.

File: composer.json [hunk 8]
```diff
@@ -22,8 +22,8 @@
       "en-GB": "FINDOLOGIC Search & Navigation"
     },
     "description": {
-      "de-DE": "FINDOLOGIC Plugin für das Shopware 6 E-Commerce System",
-      "en-GB": "FINDOLOGIC plugin for Shopware 6 e-commerce system"
+      "de-DE": "FINDOLOGIC Plugin für das Shopware 6 E-Commerce System. Features: Personalisierung, Suche, Navigation, User Interface, Merchandising. Shopware Technology Partner.",
```
Comment (dblaichinger) [inline composer.json:25]:
```suggestion
      "de-DE": "Findologic Plugin für Shopware 6. Features: Personalisierung, Suche, Navigation, User Interface, Merchandising. Shopware Technology Partner.",
```

File: composer.json [hunk 9]
```diff
@@ -22,8 +22,8 @@
...
+      "en-GB": "FINDOLOGIC plugin for the Shopware 6 e-commerce system. Features: Personalization, Search, Navigation, User Interface, Merchandising. Shopware Technology Partner."
```
Comment (dblaichinger) [inline composer.json:26]:
```suggestion
      "en-GB": "Findologic plugin for Shopware 6. Features: Personalization, Search, Navigation, User Interface, Merchandising. Shopware Technology Partner."
```

**Review by dblaichinger — CHANGES_REQUESTED** [review]

File: bin/release [hunk 10]
```diff
@@ -1,81 +1,92 @@
...
-    'FinSearch/phpcs.xml',
-    'FinSearch/phpunit.legacy.xml.dist',
-    'FinSearch/phpunit.xml.dist',
-    'FinSearch/tests/\*',
-    '*.zip',
-];
 
 $logger = new Logger('release');
-$formatter = new LineFormatter(null, null, false, true);
+$formatter = new ColoredLineFormatter(null, null, null, false, true);
```
[the remark this answers is not in the document]
Comment (TheKeymaster) [reply bin/release:16]: It ain't much, it just looks a little bit different:

l

[the remark this answers is not in the document]
Comment (dblaichinger) [reply bin/release:16]: Better than nothing :)

File: bin/release [hunk 11]
```diff
@@ -1,81 +1,92 @@
...
-    exit(1);
-}
+$logger->info('Successfully released!');
 
-exec(sprintf('git checkout tags/%s', $tag));
+function askReleaseVersion(): string
+{
+    $version = readline('Version number (e.g. 1.2.3 or 1.2.3-RC3): ');
+    if (!is_string($version)) {
+        return '';
```
[the remark this answers is not in the document]
Comment (TheKeymaster) [reply bin/release:37]: Yeah makes sense.

File: bin/release (hunk 10 above)
[the remark this answers is not in the document]
Comment (TheKeymaster) [reply bin/release:16]: image

File: .github/workflows/release.yml [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+        with:
+          upload_url: ${{ steps.create_release.outputs.upload_url }}
+          asset_path: ${{ github.workspace }}/${{ env.PLUGIN_NAME }}.zip
+          asset_name: ${{ env.PLUGIN_NAME }}-${{ steps.get_version.outputs.VERSION }}.zip
+          asset_content_type: application/zip
+
+      - name: Upload Plugin to Shopware Store
+        run: php frosh-plugin-upload.phar ext:upload $(pwd)/${PLUGIN_NAME}.zip
```
[the remark this answers is not in the document]
Comment (TheKeymaster) [reply .github/workflows/release.yml:100]: This is declared in the `env`:

env

File: .github/workflows/release.yml [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+        uses: actions/upload-release-asset@v1.0.2
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+        with:
+          upload_url: ${{ steps.create_release.outputs.upload_url }}
+          asset_path: ${{ github.workspace }}/${{ env.PLUGIN_NAME }}.zip
+          asset_name: ${{ env.PLUGIN_NAME }}-${{ steps.get_version.outputs.VERSION }}.zip
+          asset_content_type: application/zip
+
+      - name: Upload Plugin to Shopware Store
```
[the remark this answers is not in the document]
Comment (TheKeymaster) [reply .github/workflows/release.yml:99]: It is the "Shopware Store", so I will leave the `S` in store still capitalized.

File: bin/build-release (hunk 2 above)
Comment (TheKeymaster) [reply bin/build-release:61]: These will only change anyway as soon as Shopware was to create `6.5` and in this case, there will be a story that makes sure we are still compatible.

File: bin/build-release (hunk 3 above)
Comment (TheKeymaster) [reply bin/build-release:64]: We can also do it here, it doesn't really matter. It was done for the release creation process, as the `composer.json`, which is checked-in, has 2 space indents.

File: bin/build-release (hunk 4 above)
Comment (TheKeymaster) [reply bin/build-release:72]: Using `TMP_DIR`, as this actually contains `/tmp/FinSearch`.

File: bin/build-release (hunk 3 above)
Comment (dblaichinger) [reply bin/build-release:64]: Since this is duplicate code, it would be even better to have it in one place

File: composer.json (hunk 7 above)
Comment (TheKeymaster) [reply composer.json:4]: Yes the FroshPluginUploader does not allow versions with prefix `v`, so we changed this.

File: release/bin/build-release [hunk 14]
```diff
@@ -0,0 +1,20 @@
...
+declare(strict_types=1);
+
+use FINDOLOGIC\FinSearch\Release\ShopwarePluginRelease;
+
+require_once __DIR__ . '/../../vendor/autoload.php';
+
+require_once __DIR__ . '/../src/ReleaseBundler.php';
+require_once __DIR__ . '/../src/ReleaseTagger.php';
+require_once __DIR__ . '/../src/ShopwarePluginRelease.php';
+require_once __DIR__ . '/../src/Utils.php';
```
Comment (dblaichinger) [inline release/bin/build-release:13]: ReleaseBundler, ReleaseTagger, Utils seems unused.

File: release/bin/release [hunk 15]
```diff
@@ -0,0 +1,20 @@
...
+declare(strict_types=1);
+
+use FINDOLOGIC\FinSearch\Release\ShopwarePluginRelease;
+
+require_once __DIR__ . '/../../vendor/autoload.php';
+
+require_once __DIR__ . '/../src/ReleaseBundler.php';
+require_once __DIR__ . '/../src/ReleaseTagger.php';
+require_once __DIR__ . '/../src/ShopwarePluginRelease.php';
+require_once __DIR__ . '/../src/Utils.php';
```
Comment (dblaichinger) [inline release/bin/release:13]: ReleaseBundler, ReleaseTagger, Utils seems unused.

File: release/src/ShopwarePluginRelease.php [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+
+    /** @var ReleaseTagger */
+    private $releaseTagger;
+
+    public function __construct(
+        LoggerInterface $logger,
+        ?ReleaseBundler $releaseBundler = null,
+        ?ReleaseTagger $releaseTagger = null
+    ) {
+        $this->logger = $logger;
```
Comment (dblaichinger) [inline release/src/ShopwarePluginRelease.php:28]: Use `$this->buildLogger()` here and just pass the name as string.

File: release/src/ShopwarePluginRelease.php [hunk 17]
```diff
@@ -0,0 +1,64 @@
...
+        $this->releaseBundler = $releaseBundler ?? new ReleaseBundler($logger);
+        $this->releaseTagger = $releaseTagger ?? new ReleaseTagger($logger);
+    }
+
+    public static function buildLogger(string $name): LoggerInterface
```
Comment (dblaichinger) [inline release/src/ShopwarePluginRelease.php:33]: can be private then.

**Review by dblaichinger — CHANGES_REQUESTED** [review]

File: release/bin/build-release (hunk 14 above)
Comment (TheKeymaster) [reply release/bin/build-release:13]: No it is not unused, it is used inside `ReleaseTagger`and `ReleaseBundler`, but those should not care about autoloading these files.

File: release/bin/release (hunk 15 above)
Comment (TheKeymaster) [reply release/bin/release:13]: No it is not unused, it is used inside `ReleaseTagger`and `ReleaseBundler`, but those should not care about autoloading these files.

File: release/src/ShopwarePluginRelease.php (hunk 16 above)
Comment (TheKeymaster) [reply release/src/ShopwarePluginRelease.php:28]: Both files `release` and `build-release` still utilize the logger. This means if we want to add more logging in the future or add more functionality, which may be outside of `ShopwarePluginRelease`, we would have no way to access the logger.

File: release/src/ShopwarePluginRelease.php (hunk 17 above)
Comment (TheKeymaster) [reply release/src/ShopwarePluginRelease.php:33]: See https://github.com/findologic/plugin-shopware-6/pull/184#discussion_r710821486

Comment (dblaichinger) [reply release/src/ShopwarePluginRelease.php:33]: As discussed, move to Utils.

**Review by dblaichinger — APPROVED** [review]

**Pull request merged**